packages/dht/src/connection/WebRTC/NodeWebRtcConnection.ts

```diff
@@ -0,0 +1,214 @@
+                } catch (err) {
+                    console.error(err)
+                }
+            } else {
+                this.close()
+            }
+        } else {
+            this.close()
+        }
+    }
+
+    setPeerDescriptor(peerDescriptor: PeerDescriptor): void {
+        this.remotePeerDescriptor = peerDescriptor
+    }
+
+    getPeerDescriptor(): PeerDescriptor | undefined {
+        return this.remotePeerDescriptor
+    }
+
+    send(data: Uint8Array): void {
+        if (this.isOpen()) {
+            this.doSend(data)
+        }
+        else {
+            this.addToBuffer(data)
+        }
+    }
+
+    sendBufferedMessages(): void {
+        while (this.buffer.length > 0) {
+            this.send(this.buffer.shift()!)
+        }
+    }
+
+    private doSend(data: Uint8Array): void {
+        this.dataChannel?.sendMessageBinary(data as Buffer)
+    }
+
+    private addToBuffer(msg: Uint8Array): void {
+        this.buffer.push(msg)
+    }
+
+    getBufferedMessages(): Uint8Array[] {
+        return this.buffer
+    }
+
+    close(): void {
+        logger.trace(`Closing Node WebRTC Connection`)
+        if (this.connectingTimeoutRef) {
+            clearTimeout(this.connectingTimeoutRef)
+        }
+        this.emit(ConnectionEvent.DISCONNECTED)
+        if (this.dataChannel) {
+            this.dataChannel.close()
+        }
+        if (this.connection) {
+            this.connection.close()
+        }
+        this.removeAllListeners()
+    }
+
+    private onDataChannel(dataChannel: DataChannel): void {
+        this.openDataChannel(dataChannel)
+        this.setupDataChannel(dataChannel)
+    }
+
+    private setupDataChannel(dataChannel: DataChannel): void {
+        dataChannel.setBufferedAmountLowThreshold(this.bufferThresholdLow)
+        dataChannel.onOpen(() => {
+            logger.trace(`dc.onOpened`)
+            this.openDataChannel(dataChannel)
+        })
+
+        dataChannel.onClosed(() => {
+            logger.trace(`dc.closed`)
+            this.close()
+        })
+
+        dataChannel.onError((err) => logger.error(err))
+
+        dataChannel.onBufferedAmountLow( () => {
+            logger.trace(`dc.onBufferedAmountLow`)
+        })
+
+        dataChannel.onMessage((msg) => {
+            logger.trace(`dc.onMessage`)
+            this.emit(ConnectionEvent.DATA, msg as Buffer)
+        })
+    }
+
+    private openDataChannel(dataChannel: DataChannel): void {
+        if (this.connectingTimeoutRef) {
+            clearTimeout(this.connectingTimeoutRef)
+        }
+        this.dataChannel = dataChannel
+        this.sendBufferedMessages()
+        logger.trace(`DataChannel opened for peer ${this.remotePeerDescriptor.peerId.toString()}`)
+        this.emit(ConnectionEvent.CONNECTED)
+
+    }
+
+    private onStateChange(state: string): void {
+        this.lastState = state
+    }
+
+    isOpen(): boolean {
+        return this.lastState === 'connected' && !!this.dataChannel
```

Comment:
Potential bug here, it could be that node-datachannel returns `Connected`?

Reply:
It seems that node-datachannel returns lower-case states after all (maybe they fixed the bug in a later varsion)